### PR TITLE
Add support for interpreter fuzzing.

### PR DESCRIPTION
**Pull request by abizjak**
This PR adds support for fuzzing of the smart contract interpreter. Details are described in the [README](https://github.com/Concordium/concordium-wasm-smart-contracts/blob/795bb4839b2535d19794f5ec857401cd7a93d88b/README.md#fuzzing-the-smart-contract-interpreter). The fuzzing is based on generating random but valid and type-correct Wasm smart contracts. The generation is done in a [fork](https://github.com/Concordium/wasm-tools/tree/mra/fuzzing) of the [wasm-smith](https://github.com/bytecodealliance/wasm-tools/tree/main/crates/wasm-smith) random Wasm program generator.

### CODE REVIEW

File: wasm-chain-integration/src/types.rs [hunk 1]
```diff
@@ -67,9 +67,9 @@ impl<Policies> ReceiveContext<Policies> {
     pub fn self_address(&self) -> &ContractAddress { &self.self_address }
 }
 
-pub(crate) fn deserial_receive_context<'a>(
-    source: &'a [u8],
-) -> ParseResult<ReceiveContext<&'a [u8]>> {
+pub(crate) fn deserial_receive_context(
+    source: &[u8],
+) -> ParseResult<ReceiveContext<&[u8]>> {
```
Comment (abizjak) [inline wasm-chain-integration/src/types.rs]: Was this change needed?

I originally put the lifetimes there to be explicit, even though they can be elided.

File: wasm-chain-integration/Cargo.toml [hunk 2]
```diff
@@ -3,20 +3,27 @@ name = "wasm-chain-integration"
...
 [features]
 # If this feature is enabled  the ffi exports are going to be produced
 # for use from Haskell.
 enable-ffi = []
 default=["enable-ffi"]
+fuzz = ["arbitrary", "wasm-smith", "wasmprinter"]
+fuzz-coverage = ["wasm-transform/fuzz-coverage"]
+
+[workspace]
+members = ["fuzz"]
```
Comment (abizjak) [inline wasm-chain-integration/Cargo.toml:17]: Does this mean that the fuzz subpackage is built regardless of the features?

File: wasm-chain-integration/fuzz/fuzz_targets/interpreter.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    validate::validate_module,
+};
+
+const ENERGY: u64 = 10_000_000;
+const CONFIG: PrintConfig = PrintConfig {
+    print_success:                    false,
+    print_failure:                    false,
+    print_module_before_interpreting: false,
+    print_failing_module:             false,
+};
```
Comment (abizjak) [inline wasm-chain-integration/fuzz/fuzz_targets/interpreter.rs:22]: And some documentation of these fields.

**Review by abizjak — COMMENTED** [review]
I have a few nitpicky comments, but otherwise this looks very good.

File: wasm-chain-integration/src/types.rs (hunk 1 above)
Comment (amaurremi) [reply wasm-chain-integration/src/types.rs]: I got a warning about them from cargo clippy I believe, should I put them back?

File: wasm-chain-integration/src/lib.rs [hunk 4]
```diff
@@ -23,10 +22,16 @@ use wasm_transform::{
     types::{ExportDescription, Module, Name},
     utils, validate,
 };
+#[cfg(feature = "fuzz")]
+pub mod fuzz;
+#[cfg(feature = "fuzz")]
+pub mod types;
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/lib.rs:28]: I am not sure I get which example you mean, but `pub use types::*` re-exports all the definitions from `::types`.
It is true you cannot use `wasm_chain_integration::types::Foo`, but you can use `wasm_chain_integration::Foo` directly.

File: wasm-chain-integration/src/types.rs (hunk 1 above)
Comment (abizjak) [reply wasm-chain-integration/src/types.rs]: It's not a big deal so do what you think is best.

But is this using the nightly clippy? I think we do have CI setup to check this, and it does not complain with the version we use.

**Comment by abizjak** [comment]
Would you please update the PR description and then merge this?

**Review by amaurremi — APPROVED** [review]
